J4_dev_manual/Creating a Smart Search plug-in.md

```diff
@@ -0,0 +1,192 @@
+        in search results.
+  - getURL
+      -   
+        Returns the non-SEF URL for an item. The default return, using
+        plg\_finder\_content for example, is
+        'index.php?option=com\_content\&view=article\&id=1'.
+  - getStateQuery
+      -   
+        This method is used to build a generic query for checking the
+        access and state of items and their parent categories. This
+        method uses the $table and $state\_field properties to select
+        the correct data.
+
+## Event methods
+
+The following methods are called via the Smart Search content plug-in in
+response to specific trigger events in Joomla.
+
+  - onFinderBeforeSave
+      -   
+        Called in response to an onContentBeforeSave event.
+  - onFinderAfterSave
+      -   
+        Called in response to an onContentAfterSave event.
+  - onFinderAfterDelete
+      -   
+        Called in response to an onContentDelete event.
+  - onFinderChangeState
+      -   
+        Called in response to an onContentChangeState event.
+  - onFinderCategoryChangeState
+      -   
+        Called in response to an onCategoryChangeState event.
+
+## Methods in more details
+
+### getListQuery
+
+This method must return a JDatabaseQuery object that will form the basis
+of a query that will return a list of objects of the given content type.
+There are constraints on this query to ensure that it is compatible with
+other methods used by the indexer. If the following conditions are not
+true then you will need to override other methods in the
+FinderIndexerAdapter class.
+
+  - The primary database table that is being indexed must have an alias
+    of "a".\[1\]
+
+### index
+
+This method is called whenever a content item needs to be indexed (or
+re-indexed). It is passed an object of type FinderIndexerResult and its
+purpose is to make adjustments to that object and to add metadata before
+handing control to FinderIndexer::index method to do the indexing
+proper.
+
+### setup
+
+This method is called prior to the index method and is used to set up
+the index process. In this method is where you would typically include
+any externally required classes to allow for proper indexing.
+
+## FinderIndexerResult properties
+
+  - url
+  - route
+  - title
+  - description
+  - published
+  - state
+  - access
+  - language
+  - publish\_start\_date
+  - publish\_end\_date
+  - extension (only used in core for com\_categories plugin)
+
+## Testing
+
+Testing a finder plugin can be a bit difficult due to the fact that it
+is typically run using an ajax process, so you don't tend to actually
+see any results from the process.
+
+As such, one way to test is to use the command line indexer.\[2\]
+
+Open a command prompt/ssh session on the server and go to the root
+directory of the website. From there, you can run the indexer and errors
+and log messages will go to the screen. `php cli/finder_indexer.php`
+
+## Notes
+
+<references/>
```

Comment:
``` suggestion

Reply:
delete line

Reply:
How do you pick out lines for suggested changes?